packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/TaskService.java

```diff
@@ -0,0 +1,719 @@
+    TaskInterface task = getTask(taskName, appId);
+    Class unversionedConsumerClass = unversionedClassForClass(consumerClass);
+
+    // Task not found.
+    if (task == null) {
+      throw new TaskNotFoundException(taskName, appId);
+    }
+
+    // Check if the consumer is an instance of given consumer class.
+    if (unversionedConsumerClass != null && !unversionedConsumerClass.isInstance(task.getConsumer())) {
+      throw new InvalidConsumerClassException(taskName);
+    }
+
+    Map<String, TaskInterface> appTasks = sTasksTable.get(appId);
+
+    if (appTasks != null) {
+      appTasks.remove(taskName);
+    }
+
+    Log.i(TAG, "Unregistering task '" + taskName + "' for app '" + appId + "'.");
+
+    task.getConsumer().didUnregister();
+    saveTasksForAppWithId(appId);
+  }
+
+  @Override
+  public void unregisterAllTasksForAppId(String appId) {
+    Map<String, TaskInterface> appTasks = sTasksTable.get(appId);
+
+    if (appTasks != null) {
+      Log.i(TAG, "Unregistering all tasks for app '" + appId + "'.");
+
+      for (TaskInterface task : appTasks.values()) {
+        task.getConsumer().didUnregister();
+      }
+
+      appTasks.clear();
+      removeAppFromConfig(appId);
+    }
+  }
+
+  @Override
+  public boolean taskHasConsumerOfClass(String taskName, String appId, Class consumerClass) {
+    TaskInterface task = getTask(taskName, appId);
+    Class unversionedConsumerClass = unversionedClassForClass(consumerClass);
+    return task != null && unversionedConsumerClass.isInstance(task.getConsumer());
+  }
+
+  @Override
+  public Bundle getTaskOptions(String taskName, String appId) {
+    TaskInterface task = getTask(taskName, appId);
+    if (task != null) {
+      return task.getOptionsBundle();
+    }
+    return null;
+  }
+
+  @Override
+  public List<Bundle> getTasksForAppId(String appId) {
+    Map<String, TaskInterface> appTasks = sTasksTable.get(appId);
+    List<Bundle> tasks = new ArrayList<>();
+
+    if (appTasks != null) {
+      for (TaskInterface task : appTasks.values()) {
+        Bundle taskBundle = new Bundle();
+
+        taskBundle.putString("taskName", task.getName());
+        taskBundle.putString("taskType", task.getConsumer().taskType());
+        taskBundle.putBundle("options", task.getOptionsBundle());
+
+        tasks.add(taskBundle);
+      }
+    }
+    return tasks;
+  }
+
+  @Override
+  public void notifyTaskFinished(String taskName, final String appId, Map<String, Object> response) {
+    String eventId = (String) response.get("eventId");
+    List<String> appEvents = sEvents.get(appId);
+
+    Log.i(TAG, "Finished task '" + taskName + "' with eventId '" + eventId + "'.");
+
+    if (appEvents != null) {
+      appEvents.remove(eventId);
+
+      if (appEvents.size() == 0) {
+        sEvents.remove(appId);
+
+        // Invalidate app record but after 2 seconds delay so we can still take batched events.
```

Comment:
Is there a better way we could wait for this even than a timeout, similar to how one of the .ts files waits for the end of the initial event loop iteration to define tasks?